File: tic_tac_toe.cpp
//C++ version tic tac toe game

#include <iostream>
#include <stdlib.h>
#include <time.h>
using namespace std;

void print_board(char* group)
{
	cout << "\t\t| " << group[0] << " | " << group[1] << " | " << group[2] << " |\n";
	cout << "\t\t|-----------|" << endl;
	cout << "\t\t| " << group[3] << " | " << group[4] << " | " << group[5] << " |\n";
	cout << "\t\t|-----------|" << endl;
	cout << "\t\t| " << group[6] << " | " << group[7] << " | " << group[8] << " |\n";
}


char* check_three(char kind, char* group)
{
	char situation[9] = {};
	int index = 0;
	int all_possible[8][3] = {{0, 1, 2}, {3, 4, 5}, {6, 7, 8}, {0, 3, 6}, {1, 4, 7}, {2, 5, 8}, {0, 4, 8}, {2, 4, 6}};
	for (int i = 0; i < 8; i++){
		int x = all_possible[i][0];
		int y = all_possible[i][1];
		int z = all_possible[i][2];
		if((group[x]==kind && group[y]==kind && group[z]==' ') || 
			(group[x]==kind && group[y]==' ' && group[z]==kind) || 
			(group[x]==' ' && group[y]==kind && group[z]==kind)){
			if (group[x] == ' '){
				situation[index] = '0' + x;		//convert int to char
				index++;
			}
			else if (group[y] == ' '){
				situation[index] = '0' + y;		//convert int to char
				index++;
			}
			else if (group[z] == ' '){
				situation[index] = '0' + z;		//convert int to char
				index++;
			}
		}
	}
	return situation;
}


int human_step(char* group)
{
	int posit;
	cout << "Your turn. Enter a number: ";
	cin >> posit;
	while (group[posit-1] != ' '){
		cout << "Please enter a valid cell number: ";
		cin >> posit;
	}
	return posit - 1;
}


int pc_step(int step, char* group)
{
	srand((unsigned)time(NULL)); 
	if (step == 0){
		int pro_step[5] = {0, 2, 4, 6, 8};
		int position = rand() % 4;
		return pro_step[position];
	}
	else if (step == 2){
        	if (group[4] == 'X'){
			int cor_step[4] = {0, 2, 6, 8};
            		for(int i = 0; i < 4; i++){
				if (group[cor_step[i]] == 'O'){
                    			int posit = 0;
                    			if (cor_step[i] == 0)
                        			posit = 8;
                    			else if (cor_step[i] == 8)
                        			posit = 0;
                    			else if (cor_step[i] == 2)
                        			posit = 6;
                   			else if (cor_step[i] == 6)
                        			posit = 2;
                    			return posit;
				}
			}
			int nearby[2] = {0, 0};
            		if (group[1] == 'O'){
                		nearby[0] = 0;
                		nearby[1] = 2;
			}
            		else if (group[3] == 'O'){
                		nearby[0] = 0;
                		nearby[1] = 6;
			}
            		else if (group[5] == '0'){
                		nearby[0] = 2;
                		nearby[1] = 8;
			}
            		else if (group[7] == '0'){
                		nearby[0] = 6;
                		nearby[1] = 8;
			}
            		int posit = nearby[rand() % 2];
            		return posit;
		}
        	else{
            		if (group[4] == ' ')
               	 		return 4;
            		int posit = 0;
            		if (group[0] == 'X')
                		posit = 8;
            		else if (group[8] == 'X')
                		posit = 0;
            		else if (group[2] == 'X')
                		posit = 6;
            		else if (group[6] == 'X')
                		posit = 2;
            		return posit;
		}
	}
    	else{
        	char* first_arr = check_three('X', group);
        	if (strlen(first_arr) != 0){
            		int posit = first_arr[0];
            		return posit;
		}
        	char* second_arr = check_three('O', group);
        	if (strlen(second_arr) != 0){
            		int posit = second_arr[0];
            		return posit;
		}
        	int third_posit = 0;
        	int third_max = -1;
        	for (int temp = 0; temp < 10; temp++){
            		if (group[temp] == ' '){
                		if (third_max == -1){
                    			third_posit = temp;
                    			third_max = 0;
				}
                		char* temp_group = group;
                		char* temp_arr = check_three('X', temp_group);
                		if (strlen(temp_arr) > third_max){
                    			third_max = strlen(temp_arr);
                    			third_posit = temp;
				}
			}
		}
        	return third_posit;
	}
}


int whose_turn(int step)
{
	if (step % 2 == 0)
		return 1;
	else
		return 2;
}


void result(char state)
{
	if (state == 'X')
		cout << "Aha, you lose!" << endl;
	else if (state == 'O')
		cout << "Wow, you win!" << endl;
	else if (state == ' ')
		cout << "Tie! Try again." << endl;
	else
		cout << "Error." << endl;
}


bool game_over(char* group)
{
	if (group[0] == group[1] && group [1] == group[2] && group[0] != ' '){
		result(group[0]);
		return true;
	}
	else if (group[3] == group[4] && group [4] == group[5] && group[3] != ' '){
		result(group[3]);
		return true;
	}
	else if (group[6] == group[7] && group [7] == group[8] && group[6] != ' '){
		result(group[6]);
		return true;
	}
	else if (group[0] == group[3] && group [3] == group[6] && group[0] != ' '){
		result(group[0]);
		return true;
	}
	else if (group[1] == group[4] && group [4] == group[7] && group[7] != ' '){
		result(group[1]);
		return true;
	}
	else if (group[2] == group[5] && group [5] == group[8] && group[2] != ' '){
		result(group[2]);
		return true;
	}
	else if (group[0] == group[4] && group [4] == group[8] && group[0] != ' '){
		result(group[0]);
		return true;
	}
	else if (group[2] == group[4] && group [4] == group[6] && group[2] != ' '){
		result(group[2]);
		return true;
	}
	else
	{
		bool is_tie = true;
		for (int i = 0; i < 9; i++)
			if (group[i] == ' ')
				is_tie = false;
		if (is_tie)
		{
			result(' ');
			return true;
		}
		else
			return false;
	}
}



int main()
{
	char group[9] = {' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '};
	int step = 0;
	cout << "Choss a cell numbered from 1 to 9 as below and play.\n";
	cout << "\t\t| 1 | 2 | 3 |" << endl;
	cout << "\t\t|-----------|" << endl;
	cout << "\t\t| 4 | 5 | 6 |" << endl;
	cout << "\t\t|-----------|" << endl;
	cout << "\t\t| 7 | 8 | 9 |" << endl;

	while(!game_over(group))
	{
		int turn = whose_turn(step);		// 1->computer, 2->human
		if (turn == 1)
		{
			group[pc_step(step, group)] = 'X';
			cout << "Computer has done." << endl;
		}
		else if (turn == 2)
		{
			group[human_step(group)] = 'O';
			cout << "Human has done. Computer's turn." << endl;
		}
		print_board(group);
		step++;
	}

	system("pause");
	return 0;
}

File: tic_tac_toe.py
#python version tic tac toe game
import random


def human_step():
    posit = input("Your turn. Enter a number: ")
    while posit not in "123456789":
        posit = input("Please enter a valid cell number: ")
    posit = int(posit) - 1
    while group[posit] != " ":
        posit = input("Please enter a valid cell number: ")
        posit = int(posit) - 1
    return posit


def pc_step():
    if step == 0:
        pro_step = [0, 2, 4, 6, 8]
        position = random.randint(0, 4)
        return pro_step[position]
    elif step == 2:
        if group[4] == "X":
            cor_step = [0, 2, 6, 8]
            for i in range(4):
                if group[cor_step[i]] == "O":
                    posit = 0
                    if cor_step[i] == 0:
                        posit = 8
                    elif cor_step[i] == 8:
                        posit = 0
                    elif cor_step[i] == 2:
                        posit = 6
                    elif cor_step[i] == 6:
                        posit = 2
                    return posit
            nearby = [0, 0]
            if group[1] == "O":
                nearby[0] = 0
                nearby[1] = 2
            elif group[3] == "O":
                nearby[0] = 0
                nearby[1] = 6
            elif group[5] == "0":
                nearby[0] = 2
                nearby[1] = 8
            elif group[7] == "0":
                nearby[0] = 6
                nearby[1] = 8
            posit = nearby[random.randint(0, 1)]
            return posit
        else:
            if group[4] == " ":
                return 4
            posit = 0
            if group[0] == "X":
                posit = 8
            elif group[8] == "X":
                posit = 0
            elif group[2] == "X":
                posit = 6
            elif group[6] == "X":
                posit = 2
            return posit
    else:
        first_arr = check_three("X", group)
        if len(first_arr) != 0:
            posit = first_arr[0]
            return posit
        second_arr = check_three("O", group)
        if len(second_arr) != 0:
            posit = second_arr[0]
            return posit
        third_posit = 0
        third_max = -1
        for temp in range(9):
            if group[temp] == " ":
                if third_max == -1:
                    third_posit = temp
                    third_max = 0
                temp_group = group.copy()
                temp_arr = check_three("X", temp_group)
                if len(temp_arr) > third_max:
                    third_max = len(temp_arr)
                    third_posit = temp
        return third_posit


def whose_turn():
    if step % 2 == 0:
        return "cp"
    if step % 2 == 1:
        return "human"


def check_three(kind, gp):
    situation = []
    all_possible = [[0, 1, 2], [3, 4, 5], [6, 7, 8], [0, 3, 6], [1, 4, 7], [2, 5, 8], [0, 4, 8], [2, 4, 6]]
    for i in range(8):
        x = all_possible[i][0]
        y = all_possible[i][1]
        z = all_possible[i][2]
        if (gp[x] == kind and gp[y] == kind and gp[z] == " ") or (gp[x] == kind and gp[y] == " " and gp[z] == kind) or (gp[x] == " " and gp[y] == kind and gp[z] == kind):
            if gp[x] == " ":
                situation.append(x)
            elif gp[y] == " ":
                situation.append(y)
            elif gp[z] == " ":
                situation.append(z)
    return situation


def result(state):
    if state == "X":
        print("Aha, you lose!")
    elif state == "O":
        print("Wow, you win!")
    elif state == " ":
        print("Tie! Try again.")


def game_over():
    if group[0] == group[1] and group[1] == group[2] and group[0] != " ":
        result(group[0])
        return True
    elif group[3] == group[4] and group[4] == group[5] and group[3] != " ":
        result(group[3])
        return True
    elif group[6] == group[7] and group[7] == group[8] and group[6] != " ":
        result(group[6])
        return True
    elif group[0] == group[3] and group[3] == group[6] and group[0] != " ":
        result(group[0])
        return True
    elif group[1] == group[4] and group[4] == group[7] and group[1] != " ":
        result(group[1])
        return True
    elif group[2] == group[5] and group[5] == group[8] and group[2] != " ":
        result(group[2])
        return True
    elif group[0] == group[4] and group[4] == group[8] and group[0] != " ":
        result(group[0])
        return True
    elif group[2] == group[4] and group[4] == group[6] and group[2] != " ":
        result(group[2])
        return True
    else:
        is_tie = True
        for i in range(9):
            if group[i] == " ":
                is_tie = False
        if is_tie:
            result(" ")
            return True
        else:
            return False


def print_board(gp):
    print("\t\t| " + str(gp[0]) + " | " + str(gp[1]) + " | " + str(gp[2]) + " |")
    print("\t\t|-----------|")
    print("\t\t| " + str(gp[3]) + " | " + str(gp[4]) + " | " + str(gp[5]) + " |")
    print("\t\t|-----------|")
    print("\t\t| " + str(gp[6]) + " | " + str(gp[7]) + " | " + str(gp[8]) + " |")


group = [" ", " ", " ", " ", " ", " ", " ", " ", " "]
step = 0
print("Choose a cell numbered from 1 to 9 as below and play.")
print("\t\t| 1 | 2 | 3 |")
print("\t\t|-----------|")
print("\t\t| 4 | 5 | 6 |")
print("\t\t|-----------|")
print("\t\t| 7 | 8 | 9 |")

while not game_over():
    turn = whose_turn()
    if turn == "human":
        group[human_step()] = "O"
        print("Human has done. Computer's turn.")
    if turn == "cp":
        group[pc_step()] = "X"
        print("Computer has done.")
    print_board(group)
    step += 1


